integrations-and-supported-tools/pytorch/scripts/Neptune_Pytorch_Support_Advanced.py

```diff
@@ -0,0 +1,182 @@
+
+    # Instatiate model, crit & opt
+    model = BaseModel(params['input_sz'], params['input_sz'], params['n_classes']).to(params['device'])
+    criterion = nn.CrossEntropyLoss()
+    optimizer = optim.SGD(model.parameters(), lr=params['lr'])
+
+    # Step 2: Log config & hyperpararameters
+    run['config/dataset/path'] = data_dir
+    run['config/dataset/transforms'] = data_tfms
+    run['config/dataset/size'] = dataset_size
+    run['config/model'] = type(model).__name__
+    run['config/criterion'] = type(criterion).__name__
+    run['config/optimizer'] = type(optimizer).__name__
+    run['config/hyperparameters'] = params
+
+    epoch_loss = 0.0
+    epoch_acc = 0.0
+    best_acc = 0.0
+
+    # Step 3: Log losses and metrics 
+    for epoch in range(params['epochs']):
+        running_loss = 0.0
+        running_corrects = 0
+        
+        for i, (x, y) in enumerate(trainloader, 0):
+            x, y = x.to(params['device']), y.to(params['device'])
+            optimizer.zero_grad()
+            outputs = model.forward(x)
+            _, preds = torch.max(outputs, 1)
+            loss = criterion(outputs, y)
+            acc = (torch.sum(preds == y.data)) / len(x)
+
+            # Log batch loss
+            run["training/batch/loss"].log(value = loss)
+
+            # Log batch accuracy
+            run["training/batch/acc"].log(value = acc)
+
+            loss.backward()
+            optimizer.step()
+        
+            running_loss += loss.item()
+            running_corrects += torch.sum(preds == y.data)
+            
+        epoch_loss = running_loss/dataset_size['train']
+        epoch_acc = running_corrects.double().item() / dataset_size['train']
+
+        # Log epoch loss
+        run[f"training/epoch/loss"].log(value = epoch_loss, step = epoch + 1)
+
+        # Log epoch accuracy
+        run[f"training/epoch/acc"].log(value = epoch_acc, step = epoch + 1)
+
+        print(f'Epoch:{epoch+1}, Loss: {epoch_loss}, Acc: {epoch_acc}')
+        if epoch_acc > best_acc:
+            best_acc = epoch_acc
+
+            # Saving model arch & weights
+            save_model(model, params['model_filename'])
+            print('Saving model -- Done!')
+
+
+    # Step 4: Log model arch & weights -- > link to adding artifacts
+    run[f"io_files/artifacts/{params['model_filename']}_arch"].upload(f"./{params['model_filename']}_arch.txt")
+    run[f"io_files/artifacts/{params['model_filename']}"].upload(f"./{params['model_filename']}.pth")
+
+
+    # Step 5: Log Images & Predictions
+    classes = ['airplane', 'automobile', 'bird', 'cat', 'deer', 'dog', 'frog','horse','ship','truck']
+    dataiter = iter(validloader)
+    images, labels = dataiter.next()
+    images_arr = []
+    labels_arr = []
+    pred_arr = []
+    model.eval()
+    # moving model to cpu for inference 
+    if torch.cuda.is_available(): model.to("cpu")
+    # iterating on the dataset to predict the output
+    for i in range(0,10):
```

Comment:
you had a helper for saving model so I think you should have a helper for this as well.

Reply:
True.

Done